### PR TITLE
In 152464352 refactor incident info edit

### PR DESCRIPTION
**Pull request by TylerOSI**
### Pivotal Story

- [Implement Redux validations and state management for **Incident Info** card 152464352](https://www.pivotaltracker.com/story/show/152464352)

### Purpose
To refactor the Incident Information Card on the Screening Page to use the new container pattern and redux (edit mode only).

### CODE REVIEW

File: app/javascript/actions/incidentInformationFormActions.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+export const RESET_INCIDENT_INFORMATION_FIELDS = 'RESET_INCIDENT_INFORMATION_FIELDS'
```
Comment (saraOSI) [inline app/javascript/actions/incidentInformationFormActions.js]: These constants should probably have FORM in the string. Helps separate them from actions that might happen on screening load, etc.

File: app/javascript/containers/screenings/IncidentInformationFormContainer.jsx [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  }
+  const onCancel = () => {
+    dispatch(resetFieldValues(screening))
+    dispatch(touchAllFields())
+    toggleShow()
+  }
+  const onChange = (fieldName, value) => dispatch(setField(fieldName, value))
+  const onBlur = (fieldName) => dispatch(touchField(fieldName))
+  return {
+    ...stateProps,
```
Comment (saraOSI) [inline app/javascript/containers/screenings/IncidentInformationFormContainer.jsx]: This is convenient, but it also means that you're sending screening and screeningWithEdits to the component, which are props it doesn't need (and, frankly, shouldn't have). Can you either find a way to pull them out, or deconstruct the stateProps object and return just the ones you need to pass to the form?

File: app/javascript/reducers/incidentInformationFormReducer.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        incident_county,
+        address,
+        location_type,
+      } = screening
+      const {
+        city,
+        state: usState,
+        street_address,
+        zip,
+      } = address || {city: '', state: '', street_address: '', zip: ''}
```
Comment (saraOSI) [inline app/javascript/reducers/incidentInformationFormReducer.js]: You shouldn't be setting defaults in the reducer, since it mutates the data returned by the API. This is a view concern, which should be handled in the selectors in the view itself.

File: app/javascript/reducers/incidentInformationFormReducer.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+export default createReducer(Map(), {
+  [FETCH_SCREENING_COMPLETE](state, {payload: {screening}, error}) {
+    if (error) {
+      return state
+    } else {
+      const {
+        incident_date,
+        incident_county,
+        address,
```
Comment (saraOSI) [inline app/javascript/reducers/incidentInformationFormReducer.js]: You can deconstruct all of your address attributes here, too. `const {address: {city, state, street_address, zip}} = screening`

File: app/javascript/reducers/incidentInformationFormReducer.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+          value: incident_county,
+          touched: false,
+        },
+        address: {
+          city: {
+            value: city,
+            touched: false,
+          },
+          state: {
+            value: usState,
```
Comment (saraOSI) [inline app/javascript/reducers/incidentInformationFormReducer.js]: Were you having issues using state as a prop name? That didn't seem to be affecting me, so I was just curious.

File: app/javascript/reducers/incidentInformationFormReducer.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        },
+        location_type: {
+          value: location_type,
+          touched: false,
+        },
+      })
+    }
+  },
+  [SET_INCIDENT_INFORMATION_FIELD](state, {payload: {field, value}}) {
+    return state.setIn(field.concat(['value']), value)
```
Comment (saraOSI) [inline app/javascript/reducers/incidentInformationFormReducer.js]: Why use `concat` here? `state.setIn([field, 'value'], value)` should work just fine, right?

File: app/javascript/reducers/incidentInformationFormReducer.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    return state.setIn([field, 'touched'], true)
+  },
+  [RESET_INCIDENT_INFORMATION_FIELDS](state, {payload: {screening}}) {
+    const {
+      incident_date,
+      incident_county,
+      address,
+      location_type,
+    } = screening
+    const {
```
Comment (saraOSI) [inline app/javascript/reducers/incidentInformationFormReducer.js]: Same comment here re: deconstructing attributes.

File: app/javascript/screenings/IncidentInformationCardView.jsx [hunk 8]
```diff
@@ -62,31 +43,15 @@ export default class IncidentInformationCardView extends React.Component {
...
-        onEdit: this.onEdit,
-        screening: this.props.screening,
-      },
-    }
-    const IncidentInformationView = (mode === 'edit') ? IncidentInformationEditView : IncidentInformationShowContainer
-    const props = allProps[mode]
+    const IncidentInformationView = (mode === 'edit') ? IncidentInformationFormContainer : IncidentInformationShowContainer
     return (
       <div className={`card ${mode} double-gap-top`} id='incident-information-card'>
         <ScreeningCardHeader
```
Comment (saraOSI) [inline app/javascript/screenings/IncidentInformationCardView.jsx]: In the other forms, we've been moving the card header inside the form/show view themselves. I'd like to get this merged in, so I think we should flag this for a further refactor.

File: app/javascript/selectors/screening/incidentInformationFormSelector.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+import {createSelector} from 'reselect'
+import {Map, List, fromJS} from 'immutable'
+import {getScreeningSelector} from 'selectors/screeningSelectors'
+import {isFutureDatetimeCreate, combineCompact} from 'utils/validator'
+
+export const getIncidentInformationFormSelector = (state) => state.get('incidentInformationForm', Map())
+
+export const getIncidentDateSelector = createSelector(
+  getIncidentInformationFormSelector,
+  (incidentInformationForm) => incidentInformationForm.getIn(['incident_date', 'value'], '')
```
Comment (saraOSI) [inline app/javascript/selectors/screening/incidentInformationFormSelector.js]: The default passed to Immutable `getIn` will only work if it returns undefined. If it finds a value of `null`, it will return that and not the default you've asked for. Depending on what you want to pass to the view, you might want to use `|| ''` instead. (This comment applies to multiple selectors in this file.)

File: app/javascript/views/IncidentInformationForm.jsx [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+      <div className='centered'>
+        <button className='btn btn-primary' onClick={onSave}>Save</button>
+        <button className='btn btn-default' onClick={onCancel}>Cancel</button>
+      </div>
+    </div>
+  </div>
+)
+
+IncidentInformationForm.propTypes = {
+  address: PropTypes.shape({
```
Comment (saraOSI) [inline app/javascript/views/IncidentInformationForm.jsx]: Our preference is to pass data elements as top-level props, as opposed to nested elements. This isn't too bad, though, so I'm cool with it. I just wanted to point it out.

File: spec/features/screening/edit_screening_spec.rb [hunk 11]
```diff
@@ -53,7 +53,7 @@
       stub_empty_relationships_for_screening(existing_screening)
       stub_empty_history_for_screening(existing_screening)
       visit edit_screening_path(id: existing_screening.id)
-      expect(page).to have_content 'Edit Screening #My Bad!'
+      expect(page.find('h1', text: 'Edit Screening #My Bad!')).to be_truthy
```
Comment (saraOSI) [inline spec/features/screening/edit_screening_spec.rb]: Why the change here?

File: app/javascript/views/IncidentInformationForm.jsx [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+    city: PropTypes.string,
+    state: PropTypes.string,
+    streetAddress: PropTypes.string,
+    zip: PropTypes.string,
+  }),
+  counties: PropTypes.arrayOf(PropTypes.shape({
+    key: PropTypes.string,
+    name: PropTypes.string,
+  })),
+  errors: PropTypes.shape({
```
Comment (saraOSI) [inline app/javascript/views/IncidentInformationForm.jsx]: Our new preference is to pass errors with the field the errors apply to, rather than one large errors array. Again, ok for now, but for future work...

File: app/javascript/reducers/incidentInformationFormReducer.js (hunk 4 above)
Comment (TylerOSI) [reply app/javascript/reducers/incidentInformationFormReducer.js]: we tried that, but it threw an error when address was nil

File: app/javascript/reducers/incidentInformationFormReducer.js (hunk 5 above)
Comment (TylerOSI) [reply app/javascript/reducers/incidentInformationFormReducer.js]: we were having issues because state is a parameter to the involved function. it cant deconstruct the object and assign state as a const because the variable state already exists.

File: app/javascript/reducers/incidentInformationFormReducer.js (hunk 6 above)
Comment (TylerOSI) [reply app/javascript/reducers/incidentInformationFormReducer.js]: field is an array because we need to handle addresses that have nested attributes, we can rename `field` if that makes it clearer.

File: spec/features/screening/edit_screening_spec.rb (hunk 11 above)
Comment (TylerOSI) [reply spec/features/screening/edit_screening_spec.rb]: when we ran the test it failed on the expectation even though the screenshot had the text. I think `have_content` doesn't wait for the page to load, and I know that visit doesn't wait. So, I added a find to make sure this test wasn't flaking.
